sprokit/src/sprokit/python/util/python_gil.h

```diff
@@ -36,37 +36,126 @@
+#include <pybind11/cast.h>
+
+#include <memory>
 
 /**
  * \file python_gil.h
  *
- * \brief RAII class for grabbing the Python GIL.
+ * \brief Helper utilities for grabbing the Python GIL.
  */
 
 namespace sprokit {
 namespace python {
 
 /**
- * \class python_gil python_gil.h <sprokit/python/util/python_gil.h>
+ * \class python_gil_cond_release python_gil.h <sprokit/python/util/python_gil.h>
  *
- * \brief Grabs the Python GIL and uses RAII to ensure it is released.
+ * \brief Releases the python gil if held by this thread for the duration this
+ * class is in scope, but only if this is a valid python thread otherwise this
+ * is a no-op.
  */
-class SPROKIT_PYTHON_UTIL_EXPORT python_gil
+class SPROKIT_PYTHON_UTIL_EXPORT python_gil_cond_release
   : private kwiver::vital::noncopyable
 {
   public:
     /**
      * \brief Constructor.
      */
-    python_gil();
+    python_gil_cond_release();
     /**
      * \brief Destructor.
      */
-    ~python_gil();
+    ~python_gil_cond_release();
+
+  private:
+    std::unique_ptr< pybind11::gil_scoped_release > rel;
+};
+
+/**
+ * \class Record settings relating to the GIL
+ */
+class SPROKIT_PYTHON_UTIL_EXPORT python_gil_settings
+{
+  public:
+    static bool cycle_gil_lock();
+    static void set_cycle_option( bool opt = true );
+
   private:
-    PyGILState_STATE const state;
+    python_gil_settings() : cycle_gil_flag( false ) {}
+    ~python_gil_settings() {}
+
+    static python_gil_settings instance;
+    bool cycle_gil_flag;
+
+  public:
+    python_gil_settings( python_gil_settings const& ) = delete;
+    void operator=( python_gil_settings const& )      = delete;
 };
 
+/**
+ * \brief Helper macros for warningless pybind11 GIL scoped acquisitions
+ */
+#define SPROKIT_SCOPED_GIL_ACQUIRE_START                                          \
+  {                                                                               \
+    pybind11::gil_scoped_acquire acquire;                                         \
+                                                                                  \
+    (void) acquire;
+
+#define SPROKIT_SCOPED_GIL_ACQUIRE_END                                            \
+  }
+
+
+#define SPROKIT_SCOPED_GIL_RELEASE_AND_ACQUIRE_START                              \
+  {                                                                               \
+    pybind11::gil_scoped_release release;                                         \
+    {                                                                             \
+      pybind11::gil_scoped_acquire acquire;                                       \
+                                                                                  \
+      (void) release;                                                             \
+      (void) acquire;
+
+#define SPROKIT_SCOPED_GIL_RELEASE_AND_ACQUIRE_END                                \
+    }                                                                             \
+  }
+
```

Comment:
This one is a little more complicated than the previous one, but it still obfuscates the real block structure. My main concern is that the ..._END macro gets placed incorrectly with respect to the surrounding block structure. A block matching editor will not help in this case. 

One possible alternate implementation could be to include the code that is in the scope as part of the macro as follows:


    #define SPROKIT_SCOPED_GIL_RELEASE_AND_ACQUIRE(CODE)                              \
      {                                                                               \
        pybind11::gil_scoped_release release;                                         \
        {                                                                             \
          pybind11::gil_scoped_acquire acquire;                                       \
          (void) release;                                                             \
          (void) acquire;
          CODE;                                                                 \
        }                                                                             \
      }